design-documents/service-isolation/authentication-authorization.md

```diff
@@ -0,0 +1,146 @@
+and ER diagram looks like this:
+
+![Current RBAC ER](current-rbac-er.png)
+
+Where `resource_id` is represented as module name + resource name (`Magento_SalesArchive::remove`, `Magento_SalesArchive::remove`, etc.),
+and each resource has permission `allow` or `deny`. But current schema does not apply to WEB API because now, each WEB API
+entry point should specify the resource. Also, the permission column looks like obvious, if a resource is not specified for
+a rule when the resource is not allowed.
+
+It's better to expose all available API and use entry points instead of existing resources as rules. For example,
+`rest/V1/shipment/:id` and HTTP verb (like POST or GET), this approach is similar to [Amazon ARNs](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html)
+and allows making fine-grained access control.
+
+As Magento should provide [multiple BFF](https://github.com/magento/architecture/blob/master/design-documents/service-isolation.md#backends-for-frontends)
+it makes sense to extract authentication and authorization flow to a separate `Auth Service` to avoid credentials
+validation and duplication of roles & privileges data. And the general schema might look like this:
+
+![Authentication and Authorization General Schema](auth-general-schema.png)
+
+1. An unauthorized user sends a request with credentials to retrieve an access token, BFF/Monolith proxies this request to
+`Auth Service` which performs authorization & authentication and issues access token based on a user role.
+2. The user sends the request to get a needed resource with the access token, BFF/Monolith proxies this request to the appropriate
+service and there are few multiple strategies for token validation:
+    - BFF/Monolith just verifies the token signature (each service sends a token for validation to `Auth Service`)
+    - BFF/Monolith sends the token to `Auth Service` for validation
+3. A service sends the token for validation to `Auth Service` and, if a request is allowed, the service processes the request.
+4. In case, if the service requires additional data from another service, it creates a new request and set received token
+as in original request. So, additional service can send the token for validation to `Auth Service` to check user permissions.
+
+The basic scenario might be as following:
+
+![Auth Basic Scenario](auth-basic-scenario.png)
+
+A token validation per each request on `Auth Service` allows keeping authorization process in the centralized place, only
+`Auth Service` knows about all users, roles and assigned privileges. Each service during deployment should expose their API
+to `Auth Service`, which can automatically assign API to existing roles based on default rules. The token validation includes
+not only signature verification but authorization.
+
+### Services Deployment
+
+`Auth Service` stores allowed operations for roles and, during deployment, each service should send a list of available
```

Comment:
We could have simple (operation performed by one service) and composite operations (operation performed by multiple services). Composite operation can be initiated by the service (for instance Checkout service calls Totals service) or BFF. How are we going to give permissions to these operations? Seems like we need to have some sort of configuration.

Reply:
Probably we need to define the list of "standard" operations for a separate service and define the privileges for them. Other permissions can be added manually.
As service owners, we know about all flows (which includes requests to other services) and can define all operations pretty accurate.